census_view.cpp

```diff
@@ -932,6 +936,844 @@ inline std::vector<std::string> const& CensusViewDataViewModel::all_headers() co
+        val.SetRange(as_range->universal_minimum(), as_range->universal_maximum());
+        SetValidator(val);
+
+        wxGridCellTextEditor::BeginEdit(row, col, grid);
+    }
+};
+
+// class RangeDateEditor
+
+class RangeDateEditor
+    :public wxGridCellDateEditor
+{
+  public:
+    RangeDateEditor() = default;
+
+    void BeginEdit(int row, int col, wxGrid* grid) override
+    {
+        auto const& value = cell_at(*grid->GetTable(), row, col);
+        auto const* as_range = member_cast<tn_range_base>(value);
+        LMI_ASSERT(typeid(calendar_date) == as_range->value_type());
+
+        auto const min_value = as_range->universal_minimum();
+        auto const max_value = as_range->universal_maximum();
+
+        DatePicker()->SetRange
+            (ConvertDateToWx(jdn_t(bourn_cast<int>(min_value)))
+            ,ConvertDateToWx(jdn_t(bourn_cast<int>(max_value)))
+            );
+
+        wxGridCellDateEditor::BeginEdit(row, col, grid);
+    }
+};
+
+class DatumSequenceEditor
+    :public wxGridCellEditor
+{
+  public:
+    DatumSequenceEditor() = default;
+    DatumSequenceEditor(DatumSequenceEditor const&) = delete;
+    DatumSequenceEditor& operator=(DatumSequenceEditor const&) = delete;
+
+    void Create(wxWindow* parent, wxWindowID id, wxEvtHandler* evtHandler) override;
+
+    void BeginEdit(int row, int col, wxGrid* grid) override;
+    bool EndEdit
+        (int row
+        ,int col
+        ,wxGrid const* grid
+        ,wxString const& oldval
+        ,wxString* newval
+        ) override;
+    void ApplyEdit(int row, int col, wxGrid* grid) override;
+
+    void Reset() override;
+
+    wxGridCellEditor* Clone() const override;
+
+    wxString GetValue() const override;
+
+  private:
+    InputSequenceEntry* Entry() const;
+
+    wxString     value_;
+    Input const* input_;
+    std::string  field_;
+};
+
+void DatumSequenceEditor::Create
+    (wxWindow*     parent
+    ,wxWindowID    id
+    ,wxEvtHandler* evtHandler
+    )
+{
+    m_control = new(wx) InputSequenceEntry(parent, id, "sequence_editor");
+
+    wxGridCellEditor::Create(parent, id, evtHandler);
+}
+
+void DatumSequenceEditor::BeginEdit(int row, int col, wxGrid* grid)
+{
+    auto table = grid->GetTable();
+    value_ = table->GetValue(row, col);
+    input_ = &row_at(*table, row);
+    field_ = col_name(*table, col);
+
+    auto entry = Entry();
+
+    entry->text_ctrl().SetValue(value_);
+    entry->input(*input_);
+    entry->field_name(field_);
+}
+
+bool DatumSequenceEditor::EndEdit
+    (int
+    ,int
+    ,wxGrid const*
+    ,wxString const&
+    ,wxString*       newval
+    )
+{
+    auto value = Entry()->text_ctrl().GetValue();
+
+    if(value == value_)
+        {
+        return false;
+        }
+
+    value_ = value;
+
+    if(newval)
+        {
+        *newval = GetValue();
+        }
+
+    return true;
+}
+
+void DatumSequenceEditor::ApplyEdit(int row, int col, wxGrid* grid)
+{
+    grid->GetTable()->SetValue(row, col, GetValue());
```

Comment:
I'd use `value_` here instead of calling `GetValue()` again. It's (slightly) more clear because `ApplyEdit()` is supposed to apply the changes stored by `EndEdit()` and `EndEdit()` updates `value_`.